pkg/agent/svid/store/service.go

```diff
@@ -0,0 +1,155 @@
+package store
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/agent/catalog"
+	"github.com/spiffe/spire/pkg/agent/manager/pipe"
+	"github.com/spiffe/spire/pkg/agent/plugin/svidstore"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+type Service interface {
+	// Run start the store service. It will block until the context is cancelled.
+	Run(ctx context.Context) error
+}
+
+type Config struct {
+	Log     logrus.FieldLogger
+	PipeOut pipe.Out
+	Catalog catalog.Catalog
+}
+
+const (
+	typeSelector = "type"
+)
+
+func New(c Config) Service {
+	return &store{
+		c: &c,
+	}
+}
+
+type store struct {
+	c *Config
+}
+
+func (p *store) Run(ctx context.Context) error {
+	err := util.RunTasks(ctx,
+		p.run,
+	)
+
+	switch {
+	case err == nil || err == context.Canceled:
+		p.c.Log.Info("Service stopped")
+		return nil
+	default:
+		p.c.Log.WithError(err).Error("Service crashed")
+		return err
+	}
+}
+
+func (p *store) run(ctx context.Context) error {
+	log := p.c.Log
+
+	// Get all configured SVID stores and create a map with them
+	// keyed by the plugin name
+	svidStores := make(map[string]svidstore.SVIDStore)
+	for _, pp := range p.c.Catalog.GetSVIDStores() {
+		svidStores[pp.Name()] = pp
+	}
+
+	// This is a defensive check. This code is only reachable when there is
+	// at least one SVID Store plugin
+	if len(svidStores) == 0 {
+		return errors.New("no SVID store provided")
+	}
+
+	for {
+		select {
+		case update := <-p.c.PipeOut.GetUpdate():
+			log := log.WithField(telemetry.RegistrationID, update.Entry.EntryId)
+			// Get SVID store name from the selectors of the entry
+			pluginName, err := getPluginName(update.Entry.Selectors)
+			if err != nil {
+				log.WithError(err).Debugf("Unable to get plugin name from selectors")
+				continue
+			}
+
+			log = log.WithField("plugin_name", pluginName)
+
+			svidStore, ok := svidStores[pluginName]
```

Comment:
was it also considered to have each SVID Store plugin have its own distinct Pipe, and giving agent a mapping to push SVIDs to the correct pipe? Then we'd have multiple Store instances running, 1 for each configured Store, and we may not have to pass along the catalog or access it as much.

I'm not sure if there are major benefits or cons to either approach, just bringing it up as an option.

Reply:
It was something I was thinking on too, but I didnt apply something like that, just because we need things to define, for example, 
`is it possible to have multiples SVID Stores per agent`? I dont know if we may restrict an SVID Store per agent (the same way we restrict attestor).
But I think if that we allow multiple, it is a great idea to have them separated, and we may filter "storable" SVIDs in cache. instead of actual approach of filter them on store service.